Feature/authordateorder

asfarley opened the pull request:
<!-- Please read CONTRIBUTING.md before submitting a pull request -->

Fixes #6826 

## Proposed changes

Add option to view commits sorted by author-date rather than commit-date. Option is accessible through View menu, and under Appearance Settings.

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![Sorted by author-date](https://i.imgur.com/HmgRDkm.png)

### After

![Sorted by commit-date](https://i.imgur.com/5o9SgCS.png)

## Test methodology 
Tested with one repository where author-dates are not the same as commit-dates. Tried enabling and disabling setting through View menu and under Appearance Settings; both methods yielded the expected result in terms of commit-ordering.

## Test environment(s) 

- GIT 2.17.1.windows.2
- Windows NT 10.0.17763.0
- .NET 4.7.3416.0

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

gerhardol commented on GitCommands/Settings/AppSettings.cs:
```diff
@@ -424,6 +424,12 @@ public static bool ShowGitStatusForArtificialCommits
 
+        public static bool SortByAuthorDate
+        {
+            get => GetBool("sortbyauthordate", true);
```
Default should ba false, much slower appearance for larger repos

asfarley replied:
Fixed.

asfarley replied:
@gerhardol I believe I've addressed all the comments - is there anything else I need to do?

RussKie commented on Build/transifex/English.Plugins.xlf:
```diff
@@ -0,0 +1,1442 @@
+<?xml version="1.0" encoding="utf-8"?>
```
❗️  There must be no changes under Build/ folder.

asfarley replied:
Fixed.

RussKie commented on GitUI/CommandsDialogs/FormBrowse.Designer.cs:
```diff
@@ -1741,13 +1744,14 @@ private void InitializeComponent()
             // FormBrowse
             // 
-            this.AutoScaleDimensions = new System.Drawing.SizeF(96F, 96F);
+            this.AutoScaleDimensions = new System.Drawing.SizeF(144F, 144F);
```
❗️  All UI changes must be done at 100% scale factor.

VS typically prompts to do that
![image](https://user-images.githubusercontent.com/4403806/59270744-9c314080-8c95-11e9-81a8-dd51ae677334.png)


Please run VS with `/noscale` switch 
![image](https://user-images.githubusercontent.com/4403806/59431442-4d67e000-8e28-11e9-91e0-029b2dbc8e9f.png)

RussKie replied:
❗️  All UI changes must be done at 100% scale factor.

Please refer to https://github.com/gitextensions/gitextensions/wiki/Guidelines%3A-UI-design

RussKie commented on GitUI/CommandsDialogs/FormBrowse.Designer.cs:
```diff
@@ -37,22 +37,22 @@ private void InitializeComponent()
             this._NO_TRANSLATE_WorkingDir = new System.Windows.Forms.ToolStripSplitButton();
```
Please explain the nature of changes to this file

RussKie commented on GitUI/CommandsDialogs/FormBrowse.cs:
```diff
@@ -341,6 +341,7 @@ void InitCountArtificial(out GitStatusMonitor gitStatusMonitor)
 
                     bool countToolbar = AppSettings.ShowGitStatusInBrowseToolbar;
                     bool countArtificial = AppSettings.ShowGitStatusForArtificialCommits && AppSettings.RevisionGraphShowWorkingDirChanges;
+                    bool byAuthorDate = AppSettings.SortByAuthorDate;
```
Where is this used?

asfarley replied:
This was an oversight, fixed.

RussKie commented on GitUI/Translation/English.xlf:
```diff
@@ -111,10 +111,6 @@ This action will be performed without warning while checking out branch.</source
-      <trans-unit id="$this.Text">
-        <source>Appearance</source>
-        <target />
-      </trans-unit>
```
This must not change. Please rebase on to the latest master

pmiossec replied:
Revert this change to make the build pass.

asfarley replied:
Fixed.

asfarley commented on GitUI/CommandsDialogs/FormBrowse.Designer.cs:
```diff
@@ -1741,13 +1744,14 @@ private void InitializeComponent()
             // FormBrowse
             // 
-            this.AutoScaleDimensions = new System.Drawing.SizeF(96F, 96F);
+            this.AutoScaleDimensions = new System.Drawing.SizeF(144F, 144F);
```
Fixed.

asfarley commented:
@pmiossec The AppVeyor build is failing again. 

I originally submitted a branch where the translation files had not been touched, but I then checked the AppVeyor output, and it seems like it was telling me that I *should* be running some tool to update those files. After running the translation tools for Debug and Release branches, the AppVeyor build started to pass. So it seems like I was seeing the opposite result of what you're suggesting.

Am I misunderstanding something?

pmiossec commented:
The artifact of your failing build tell you that the translation file 'English.xlf'  is not OK.

https://ci.appveyor.com/project/gitextensions/gitextensions/builds/25890898/artifacts

You have committed the remove of some lines (probably an error done by the script) that should be kept.

So revert just the part of this file and that should be good

asfarley commented:
@pmiossec  Sorry, I don't quite understand. In my latest squashed commit, it appears that I have not changed English.xlf at all. 

Looking under the Build/ folder, English.xlf does not exist, same as the previous commit on the master-branch. Can you clarify?

pmiossec commented:
>@pmiossec Sorry, I don't quite understand. In my latest squashed commit, it appears that I have not changed English.xlf at all

Oh, yes, indeed. I don't know why I was looking the not up to date github page and consequently also not the good build result. 

So, now, you problem is that you reverted all the xlf file instead of just the deleted lines. So you have to update the xlf (by running the translation script) like the build artifact tell you that translation is missing :

https://ci.appveyor.com/project/gitextensions/gitextensions/builds/25890898/artifacts

But be careful to commit just the lines corresponding to your development.

asfarley commented:
@pmiossec  @RussKie  I think I've applied all your comments, thanks for helping this through.

pmiossec commented:
@asfarley Forget my last review message, I did it myself.

Thanks.

pmiossec commented:
So, the new menu looks like: 
![image](https://user-images.githubusercontent.com/460196/61121620-ed547000-a49f-11e9-9bc6-74dacc812cb2.png)

RussKie commented:
Perhaps the menu should be

Sort revisions...
  > Commit date
  > Author date

On Fri, 12 Jul 2019 at 21:03, Philippe Miossec <notifications@github.com>
wrote:

> So, the new menu looks like:
> [image: image]
> <https://user-images.githubusercontent.com/460196/61121620-ed547000-a49f-11e9-9bc6-74dacc812cb2.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6854?email_source=notifications&email_token=ABBTEXQGUSAAVRXG6FDECWLP7BQJTA5CNFSM4H3DXTPKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODZZOI4Y#issuecomment-510846067>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBTEXT3V43Y6SNJXLJ64LTP7BQJTANCNFSM4H3DXTPA>
> .
>

pmiossec commented:
Yes, that's better. I let @asfarley do it...

RussKie commented:
Please submit a PR to the doco repo detailing the change and associated
side effects.
Also suggest adding a tooltip hinting performance degradation.

@gerhardol please hold on to merging the PR until the above is done.

Thank you

On Sat, Jul 13, 2019, 5:42 AM Gerhard Olsson <notifications@github.com>
wrote:

> *@gerhardol* commented on this pull request.
>
> The direct comments have been addressed, this looks OK.
> However, this still affects the perceived performance.
> Without --author-date-order GE starts displaying commits directly when
> git-log runs. With the option, git-log has to complete listing all commits
> before the command provides data to GE.
> For the GE repo, git-log requires less than one second for me, so not much
> difference. For the Linux repo git-log requires about 30s for me, so the GE
> display is white with a spinner for 30s with this option.
>
> This option works fine for most repos so the question is how to document
> the functionality?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gitextensions/gitextensions/pull/6854?email_source=notifications&email_token=ABBTEXXGBH2CSVPGZ2YSQFLP7DNDJA5CNFSM4H3DXTPKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB6KU2UQ#pullrequestreview-261442898>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABBTEXU2B3ZGU6PDNTDCXYDP7DNDJANCNFSM4H3DXTPA>
> .
>

gerhardol commented:
I do not want to block merge of something is useful, but still want users to acknowledge what they are doing.

A tooltip in the settings is probably the way this can be documented.
(Alternative in Settings-Performance, but that would be in addition to View.)

The documentation strictly follows the application and View is not really described. 
https://git-extensions-documentation.readthedocs.io/en/latest/browse_repository.html#view-revision-graph
But something should be added to the docs.

asfarley commented:
Hi folks, something strange happened on my feature branch; I went to rebase my latest changes one last time and force-push to my development branch, but I was unable to do so. Git gave me some warning about a stale branch. 

I ended up pulling just to see what had happened on the branch, and I saw a bunch of commits on my development branch which appeared to be coming from other GitExtensions developers. 

I don't understand how that could have happened, but I have completely deleted the development branch in my repository and recreated it. It seems that deleting the pull-request branch has caused this pull request to become closed, so I will submit one more new pull request. 

I think I've covered everything from the 2nd round of suggestions: moved the setting to a different location in the menu, added a tool-tip, and updated the docs. I will submit a pull request to the docs repo shortly.

pmiossec commented:
>Hi folks, something strange happened on my feature branch; I went to rebase my latest changes one last time and force-push to my development branch, but I was unable to do so. Git gave me some warning about a stale branch.

I should not have been enough clear but that was my fault.

I told you with:

>@asfarley Forget my last review message, I did it myself.
>Thanks.

and https://github.com/gitextensions/gitextensions/pull/6854#issuecomment-510846067
that I did the changes and pushed in your branch. I thought it would have been quicker to do it than explaining in detail. I did it because I though it would have been mergeable and quickly merged but that was before other comments from @RussKie and @gerhardol 😭

You would just have to sync your branch with the commit I pushed.

So, in the end, it appears that did the changes on your side. That's ok.

>I ended up pulling just to see what had happened on the branch, and I saw a bunch of commits on my development branch which appeared to be coming from other GitExtensions developers.

That's normal because I have synced your branch with the last commit in `upstream/master`.
Because, you undo that, your feature branch is now based on a very old commit. 
Please rebase on the last commit from `master` from the GitExtensions repository. The history will be much more readable like that...

There are also 2 things that should be done I think:

* Display the tooltip on performance also in the `View` menu 
* Do what is describe in the comment https://github.com/gitextensions/gitextensions/pull/6854#issuecomment-510878655 I'm more for the use of "commits" word instead or "revision"

And could you, please, post a screenshot of the view menu once done?
